code/__HELPERS/path.dm

```diff
@@ -0,0 +1,379 @@
+	path = pathfind_datum.start_search()
+	qdel(pathfind_datum)
+
+	SSpathfinder.mobs.found(l)
+	if(!path)
+		path = list()
+	return path
+
+/**
+ * A helper macro to see if it's possible to step from the first turf into the second one, minding things like door access and directional windows.
+ * Note that this can only be used inside the [datum/pathfind][pathfind datum] since it uses variables from said datum
+ * If you really want to optimize things, optimize this, cuz this gets called a lot
+ */
+#define CAN_STEP(cur_turf, next) (next && !next.density && cur_turf.Adjacent(next) && !(simulated_only && SSpathfinder.space_type_cache[next.type]) && !cur_turf.LinkBlockedWithAccess(next,caller, id) && (next != avoid))
+/// Another helper macro for JPS, for telling when a node has forced neighbors that need expanding
+#define STEP_NOT_HERE_BUT_THERE(cur_turf, dirA, dirB) ((!CAN_STEP(cur_turf, get_step(cur_turf, dirA)) && CAN_STEP(cur_turf, get_step(cur_turf, dirB))))
+
+/// Enumerator for the starting turf's sources value, so we know when we've hit the beginning when unwinding at the end
+#define PATH_START	-1
+
+
+/// The JPS Node datum represents a turf that we find interesting enough to add to the open list and possibly search for new tiles from
+/datum/jps_node
+	/// The turf associated with this node
+	var/turf/tile
+	/// The node we just came from
+	var/datum/jps_node/previous_node
+	/// The A* node weight (f_value = number_of_tiles + heuristic)
+	var/f_value
+	/// The A* node heuristic (a rough estimate of how far we are from the goal)
+	var/heuristic
+	/// How many steps it's taken to get here from the start (currently pulling double duty as steps taken & cost to get here, since all moves incl diagonals cost 1 rn)
+	var/number_tiles
+	/// How many steps it took to get here from the last node
+	var/jumps
+	/// Nodes store the endgoal so they can process their heuristic without a reference to the pathfind datum
+	var/turf/node_goal
+
+/datum/jps_node/New(turf/our_tile, datum/jps_node/previous_node, jumps_taken, turf/incoming_goal)
+	tile = our_tile
+	previous_node = previous_node
+	jumps = jumps_taken
+	if(previous_node)
+		number_tiles = previous_node.number_tiles + jumps
+		node_goal = previous_node.node_goal
+	else
+		number_tiles = 0
+		node_goal = incoming_goal
+
+	heuristic = get_dist(tile, node_goal)
+	f_value = number_tiles + heuristic
+
+/// The datum used to handle the JPS pathfinding, completely self-contained
+/datum/pathfind
+	/// The thing that we're actually trying to path for
+	var/atom/movable/caller
+	/// The turf where we started at
+	var/turf/start
+	/// The turf we're trying to path to (note that this won't track a moving target)
+	var/turf/end
+	/// The open list/stack we pop nodes out from
+	var/datum/heap/path/open
+	/// An assoc list that matches turfs (the key) to their nodes if said turf has one
+	var/list/openc
```

Comment:
What is `openc`? `open_nodes_to_check`? `open_nodes_checked`?
Some clarity here would be positive.